lib/std/zig/system.zig

```diff
@@ -221,7 +221,44 @@ pub const NativeTargetInfo = struct {
                     }
                 },
                 .windows => {
-                    // TODO Detect native operating system version.
+                    var version_info: std.os.windows.RTL_OSVERSIONINFOW = undefined;
+                    version_info.dwOSVersionInfoSize = @sizeOf(@TypeOf(version_info));
+
+                    switch (std.os.windows.ntdll.RtlGetVersion(&version_info)) {
+                        .SUCCESS => {},
+                        else => unreachable,
+                    }
+
+                    // Starting from the system infos build a NTDDI-like version
+                    // constant whose format is:
+                    //   B0 B1 B2 B3
+                    //   `---` `` ``--> Sub-version (Starting from Windows 10 onwards)
+                    //     \    `--> Service pack (Always zero in the constants defined)
+                    //      `--> OS version (Major & minor)
+                    const os_ver: u16 = //
+                        @intCast(u16, version_info.dwMajorVersion & 0xff) << 8 |
+                        @intCast(u16, version_info.dwMinorVersion & 0xff);
+                    const sp_ver: u8 = 0;
+                    const sub_ver: u8 = if (os_ver >= 0xA000) subver: {
+                        // There's no other way to obtain this info beside
+                        // checking the build number against a known set of
+                        // values
+                        const known_build_numbers = [_]u32{
+                            10240, 10586, 14393, 15063, 16299, 17134, 17763,
+                            18362, 18363,
+                        };
+                        for (known_build_numbers) |build, i| {
+                            if (version_info.dwBuildNumber < build)
+                                break :subver @truncate(u8, i);
+                        }
```

Comment:
When `dwBuildNumber` is for example 18362 this code would actually return 18363 (because of less than check).

However there is one more issue when build number is bigger than some known value but less than next known value, e.g. 17132 < 17140 (example) < 17763. It would assume that version is 17763, but the result probably should be 17132 instead (since some apis may not exist at this point). That is super rare case, since those pre-release builds of windows have expired long time ago, but if someone faked the clock an tried to run this code in that version of windows it would give an unreliable result.

Reply:
Oh thanks for the hint, I didn't think of that. I've pushed a new commit that rounds down to the earlier release number.